heron/healthmgr/src/java/org/apache/heron/healthmgr/common/PhysicalPlanProvider.java

```diff
@@ -19,70 +19,129 @@
   private static final Logger LOG = Logger.getLogger(PhysicalPlanProvider.class.getName());
 
   private final SchedulerStateManagerAdaptor stateManagerAdaptor;
   private final String topologyName;
 
-  private PhysicalPlan physicalPlan;
+  private PhysicalPlan cachedPhysicalPlan = null;
 
   @Inject
   public PhysicalPlanProvider(SchedulerStateManagerAdaptor stateManagerAdaptor,
-      EventManager eventManager, @Named(CONF_TOPOLOGY_NAME) String topologyName) {
+                              @Named(CONF_TOPOLOGY_NAME) String topologyName) {
     this.stateManagerAdaptor = stateManagerAdaptor;
     this.topologyName = topologyName;
-    eventManager.addEventListener(TopologyUpdate.class, new EventHandler<TopologyUpdate>() {
-      /**
-       * Invalidates cached physical plan on receiving topology update notification
-       */
-      @Override
-      public synchronized void onEvent(TopologyUpdate event) {
-        LOG.info(
-            "Received topology update event, invalidating cached PhysicalPlan: " + event.type());
-        physicalPlan = null;
-      }
-    });
-    eventManager.addEventListener(ContainerRestart.class, new EventHandler<ContainerRestart>() {
-      /**
-       * Invalidates cached physical plan on receiving container restart notification
-       */
-      @Override
-      public synchronized void onEvent(ContainerRestart event) {
-        LOG.info("Received container restart event, invalidating cached PhysicalPlan: "
-            + event.type());
-        physicalPlan = null;
-      }
-    });
   }
 
   @Override
   public synchronized PhysicalPlan get() {
-    physicalPlan = stateManagerAdaptor.getPhysicalPlan(topologyName);
-    if (physicalPlan == null) {
+    TopologyMaster.TMasterLocation tMasterLocation
+        = stateManagerAdaptor.getTMasterLocation(topologyName);
+    String host = tMasterLocation.getHost();
+    int port = tMasterLocation.getControllerPort();
+
+    // construct metric cache stat url
+    String url = "http://" + host + ":" + port + "/get_current_physical_plan";
+    LOG.fine("tmaster physical plan query endpoint: " + url);
+
+    // http communication
+    HttpURLConnection con = NetworkUtils.getHttpConnection(url);
+    NetworkUtils.sendHttpGetRequest(con);
+    byte[] responseData = NetworkUtils.readHttpResponse(con);
+    // byte to base64 string
+    String encodedString = new String(responseData);
+    LOG.fine("tmaster returns physical plan in base64 str: " + encodedString);
+    // base64 string to proto bytes
+    byte[] decodedBytes = Base64.getDecoder().decode(encodedString);
+    // construct proto obj from bytes
+    PhysicalPlan physicalPlan = null;
+    try {
+      physicalPlan = PhysicalPlan.parseFrom(decodedBytes);
+    } catch (Exception e) {
       throw new InvalidStateException(topologyName, "Failed to fetch the physical plan");
     }
+
+    cachedPhysicalPlan = physicalPlan;
     return physicalPlan;
   }
 
+  public PhysicalPlan getCachedPhysicalPlan() {
+    try {
```

Comment:
should this be `if (cachedPPlan == null) { cachedPPlan = get();} return cachedPPlan;`? Otherwise every time a completely new physical plan is fetched.

Reply:
intended. the healthmgr updates runtime config by pulling new pplan at the begin of each policy loop round.